spec/ddtrace/pin_spec.rb

```diff
@@ -94,32 +54,104 @@ def datadog_pin=(_pin)
     end
   end
 
+  describe '.set_on' do
+    subject(:set_on) { described_class.set_on(target, **options) }
+    let(:options) { { a_setting: :a_value } }
+
+    context 'given an object without a Pin' do
+      it { is_expected.to be_a_kind_of(described_class) }
+      it { expect(set_on[:a_setting]).to be(:a_value) }
+    end
+
+    context 'given an object with a Pin' do
+      context 'and a setting that does not conflict' do
+        let(:original_options) { { original_setting: :original_value } }
+
+        before { described_class.set_on(target, **original_options) }
+
+        it { is_expected.to be_a_kind_of(described_class) }
+        it { expect(set_on[:original_setting]).to be(:original_value) }
+        it { expect(set_on[:a_setting]).to be(:a_value) }
+      end
+
+      context 'and a setting that conflicts' do
+        let(:original_options) { { a_setting: :original_value } }
+
+        before { described_class.set_on(target, **original_options) }
+
+        it { is_expected.to be_a_kind_of(described_class) }
+        it { expect(set_on[:a_setting]).to be(:a_value) }
+      end
+
+      context 'which has a custom pin has already defined' do
+        let(:target_class) do
+          Class.new do
+            def datadog_pin
+              @custom_attribute
+            end
+
+            def datadog_pin=(_pin)
+              @custom_attribute = { custom_pin: true }
+            end
+          end
+        end
+
+        let(:target) { target_class.new }
+
+        before { pin.onto(target) }
+
+        it 'returns the custom pin' do
+          is_expected.to eq({ custom_pin: true }.merge(options))
+        end
+      end
+    end
+  end
+
+  describe '#initialize' do
+    before { pin }
+
+    context 'when given some options' do
+      let(:options) do
+        {
+          app: double('app'),
+          app_type: double('app_type'),
+          config: double('config'),
+          tags: double('tags'),
+        }
+      end
+
+      it do
+        expect(pin[:app]).to eq(options[:app])
+        expect(pin[:app_type]).to eq(options[:app_type])
+        expect(pin[:tags]).to eq(options[:tags])
+      end
+    end
```

Comment:
`config` seems to be missing (perhaps it's just easier to remove it from the options)